web/src/stream_data.ts

```diff
@@ -505,6 +505,34 @@ export function has_metadata_access(sub: StreamSubscription): boolean {
     }
 
     return false;
+};
+
+export function rewire_has_metadata_access(value: typeof has_metadata_access): void {
+    has_metadata_access = value;
+}
+
+export function has_content_access(sub: StreamSubscription): boolean {
+    if (!has_metadata_access(sub)) {
+        return false;
+    }
+
+    const can_add_subscribers = user_groups.is_user_in_setting_group(
+        sub.can_add_subscribers_group,
+        people.my_current_user_id(),
+    );
+    if (can_add_subscribers) {
+        return true;
+    }
+
+    if (sub.invite_only && !sub.subscribed) {
+        return false;
+    }
```

Comment:
I think this might read better if we had `if sub.subscribed` at the top returning `true`, rather than linking it with this conditional.